src/main/java/uk/gov/hmcts/darts/event/service/impl/RemoveDuplicateEventsProcessorImpl.java

```diff
@@ -0,0 +1,103 @@
+package uk.gov.hmcts.darts.event.service.impl;
+
+import jakarta.transaction.Transactional;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.stereotype.Service;
+import uk.gov.hmcts.darts.common.entity.EventEntity;
+import uk.gov.hmcts.darts.common.helper.CurrentTimeHelper;
+import uk.gov.hmcts.darts.common.repository.CaseManagementRetentionRepository;
+import uk.gov.hmcts.darts.common.repository.CaseRetentionRepository;
+import uk.gov.hmcts.darts.common.repository.EventRepository;
+import uk.gov.hmcts.darts.event.service.RemoveDuplicateEventsProcessor;
+
+import java.time.LocalDate;
+import java.time.OffsetDateTime;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Spliterator;
+import java.util.function.Function;
+
+import static java.time.ZoneOffset.UTC;
+import static java.util.Comparator.comparing;
+import static java.util.stream.Collectors.groupingBy;
+import static java.util.stream.Collectors.toList;
+import static java.util.stream.StreamSupport.stream;
+
+@Service
+@Slf4j
+public class RemoveDuplicateEventsProcessorImpl implements RemoveDuplicateEventsProcessor {
+
+    private static final int CHUNK_SIZE = 1000;
+
+    private final OffsetDateTime earliestRemovableEventDate;
+    private final int clearUpWindow;
+    private final EventRepository eventRepository;
+    private final CurrentTimeHelper currentTimeHelper;
+    private final CaseManagementRetentionRepository caseManagementRetentionRepository;
+    private final CaseRetentionRepository caseRetentionRepository;
+
+    public RemoveDuplicateEventsProcessorImpl(
+        @Value("${darts.events.duplicates.earliest-removable-event-date}") LocalDate earliestRemovableEventDate,
+        @Value("${darts.events.duplicates.clear-up-window}") int clearUpWindow,
+        EventRepository eventRepository,
+        CurrentTimeHelper currentTimeHelper,
+        CaseManagementRetentionRepository caseManagementRetentionRepository,
+        CaseRetentionRepository caseRetentionRepository) {
+
+        this.earliestRemovableEventDate = earliestRemovableEventDate
+            .atStartOfDay()
+            .atOffset(UTC);
+        this.clearUpWindow = clearUpWindow;
+        this.eventRepository = eventRepository;
+        this.currentTimeHelper = currentTimeHelper;
+        this.caseManagementRetentionRepository = caseManagementRetentionRepository;
+        this.caseRetentionRepository = caseRetentionRepository;
+    }
+
+    @Override
+    @Transactional
+    public void processRemoveDuplicateEvents() {
+        var today = currentTimeHelper.currentLocalDate();
+        var startDateTime = today.minusDays(clearUpWindow).atStartOfDay().atOffset(UTC);
+        if (startDateTime.isBefore(earliestRemovableEventDate)) {
+            startDateTime = earliestRemovableEventDate;
+        }
+        var endDateTime = today.atTime(23, 59).atOffset(UTC);
+        var duplicateEvents = new ArrayList<EventEntity>();
```

Comment:
I'm wondering if the approach of taking the `startDateTime` as `00:00:00` has some consequences if `clear-up-window` is ever configured to `0`.

The task would execute at 22:00, capturing all events between 00:00 and 23:59 that day. Practically there should be no events (?) between 22:00 and 23:59, as they'd be in the future, so really we're getting events between 00:00 and 22:00.

Events would continue to flow into the system between 22:00 and 23:59, but the next invocation of the task (the next day) wouldn't capture them. So our cleanup would only be partially cleaning up the day's events.

I think if `clear-up-window` is > 0, then the problem is solved because the previous day will captured in it's entirety, which will scoop up any missed events.

Do we need some validation to ensure `clear-up-window` is > 0?

Reply:
Good point, will have think about this.